components/eamxx/src/diagnostics/horiz_avg.cpp

```diff
@@ -0,0 +1,165 @@
+                   " - field name: " +
+                       fid.name() +
+                       "\n"
+                       " - field layout: " +
+                       layout.to_string() + "\n");
+  EKAT_REQUIRE_MSG(layout.tags()[0] == COL,
+                   "Error! HorizAvgDiag diagnostic expects a layout starting "
+                   "with the 'COL' tag.\n"
+                   " - field name  : " +
+                       fid.name() +
+                       "\n"
+                       " - field layout: " +
+                       layout.to_string() + "\n");
+
+  FieldIdentifier d_fid(m_diag_name, layout.clone().strip_dim(COL),
+                        fid.get_units(), fid.get_grid_name());
+  m_diagnostic_output = Field(d_fid);
+  m_diagnostic_output.allocate_view();
+}
+
+void HorizAvgDiag::compute_diagnostic_impl() {
+  using RangePolicy = Kokkos::RangePolicy<Field::device_t::execution_space>;
+  using TeamPolicy  = Kokkos::TeamPolicy<Field::device_t::execution_space>;
+  using TeamMember  = typename TeamPolicy::member_type;
+  using ESU         = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+
+  const auto &f = get_fields_in().front();
+  const auto &d = m_diagnostic_output;
+
+  const auto &layout = f.get_header().get_identifier().get_layout();
+  int dim0           = layout.dim(0);
+
+  d.deep_copy(0);
+
+  // Get the area field
+  const auto a = m_area;
+  // calculate total area
+  Real atot = 0.0;
+  Kokkos::parallel_reduce(
+      "HorizAvgDiag::compute_diagnostic_impl::total_area", dim0,
+      KOKKOS_LAMBDA(const int icol, Real &accum) { accum += a(icol); }, atot);
+  // ensure atot is no zero; if it is zero (e.g., point grid?) make it 1
+  atot = atot == 0.0 ? 1.0 : atot;
+
+  switch(layout.rank()) {
+    case 1: {
+      auto f_view = f.get_view<const Real *>();
+      auto d_view = d.get_view<Real>();
+
+      auto p = ESU::get_default_team_policy(1, dim0);
+      Kokkos::parallel_for(
+          d.name(), p, KOKKOS_LAMBDA(const TeamMember &m) {
+            Real sum = 0.0;
+            Kokkos::parallel_reduce(
+                Kokkos::TeamThreadRange(m, dim0),
+                [&](const int icol, Real &accum) {
+                  accum += (a(icol) / atot) * f_view(icol);
+                },
+                sum);
+            Kokkos::single(Kokkos::PerTeam(m), [&]() { d_view() = sum; });
+          });
+    } break;
+    case 2: {
+      auto f_view = f.get_view<const Real **>();
+      auto d_view = d.get_view<Real *>();
+
+      const int dim1 = layout.dim(1);
+      auto p         = ESU::get_default_team_policy(dim1, dim0);
+      Kokkos::parallel_for(
+          d.name(), p, KOKKOS_LAMBDA(const TeamMember &m) {
+            const int j = m.league_rank();
+            Kokkos::parallel_reduce(
+                Kokkos::TeamVectorRange(m, dim0),
+                [&](int icol, Real &accum) {
+                  accum += (a(icol) / atot) * f_view(icol, j);
```

Comment:
And if you compute atot at init time, you may as well store a copy of a that is already rescaled by atot.

Reply:
We already have utils for a field norm (check field_utils). I would add `dot` as a free function there as well. And regarding dot-vs-normalized_dot, I would just do dot. Normalization can be done before/after if need be. If you really want to, you can do a weighted dot product, where you do the dot between `x` and the elem-wise product of `y` with a weight vector `w`. But that becomes complicated, since in your case `w` has a different layout than `y`, and we need to talk about "which dim is the dot prod on".

This can get very confusing and misleading quickly. Imho, keep it simple. If you want to deal with weigths, you can make a _very verbose_ signature that makes it clear we're not doing a generic utility. Something like `void horiz_rescale (Field& x, const Field& w)` which for each horiz dof does `x(icol, : ) *= w(icol)`. But you have to consider "how many times will we use this?"...